open_spiel/games/phantom_go.cc

```diff
@@ -0,0 +1,301 @@
+}
+
+std::vector<Action> PhantomGoState::LegalActions() const {
+  std::vector<Action> actions{};
+  if (IsTerminal()) return actions;
+  for (VirtualPoint p : BoardPoints(board_.board_size())) {
+    if (board_.IsLegalMove(p, to_play_)) {
+      actions.push_back(board_.VirtualActionToAction(p));
+    }
+  }
+  actions.push_back(board_.pass_action());
+  return actions;
+}
+
+
+std::string PhantomGoState::ActionToString(Player player, Action action) const {
+  return absl::StrCat(
+      GoColorToString(static_cast<GoColor>(player)), " ",
+      VirtualPointToString(board_.ActionToVirtualAction(action)));
+}
+
+char GoColorToChar(GoColor c) {
+    switch (c) {
+    case GoColor::kBlack:
+        return 'X';
+    case GoColor::kWhite:
+        return 'O';
+    case GoColor::kEmpty:
+        return '+';
+    case GoColor::kGuard:
+        return '#';
+    default:
+        SpielFatalError(absl::StrCat("Unknown color ", c, " in GoColorToChar."));
+        return '!';
+    }
+}
+
+std::string PhantomGoState::ToString() const {
+  std::stringstream ss;
+  std::array<int, 2> stoneCount = board_.getStoneCount();
+  ss << "GoState(komi=" << komi_ << ", to_play=" << GoColorToString(to_play_)
+     << ", history.size()=" << history_.size() << ", " 
+     << "stones_count: w" << stoneCount[1] << " b" << stoneCount[0] << ")\n";
+  
+  ss << board_;
+
+  ss << board_.observationToString();
+
+  return ss.str();
+}
+
+bool PhantomGoState::IsTerminal() const {
+  if (history_.size() < 2) return false;
+  return (history_.size() >= max_game_length_) || superko_ ||
+         (history_[history_.size() - 1].action == board_.pass_action() &&
+          history_[history_.size() - 2].action == board_.pass_action());
+}
+
+std::vector<double> PhantomGoState::Returns() const {
+  if (!IsTerminal()) return {0.0, 0.0};
+
+  if (superko_) {
+    // Superko rules (https://senseis.xmp.net/?Superko) are complex and vary
+    // between rulesets.
+    // For simplicity and because superkos are very rare, we just treat them as
+    // a draw.
+    return {DrawUtility(), DrawUtility()};
+  }
+
+  // Score with Tromp-Taylor.
+  float black_score = TrompTaylorScore(board_, komi_, handicap_);
+
+  std::vector<double> returns(phantom_go::NumPlayers());
+  if (black_score > 0) {
+    returns[ColorToPlayer(GoColor::kBlack)] = WinUtility();
+    returns[ColorToPlayer(GoColor::kWhite)] = LossUtility();
+  } else if (black_score < 0) {
+    returns[ColorToPlayer(GoColor::kBlack)] = LossUtility();
+    returns[ColorToPlayer(GoColor::kWhite)] = WinUtility();
+  } else {
+    returns[ColorToPlayer(GoColor::kBlack)] = DrawUtility();
+    returns[ColorToPlayer(GoColor::kWhite)] = DrawUtility();
+  }
+  return returns;
+}
+
+std::unique_ptr<State> PhantomGoState::Clone() const {
+  return std::unique_ptr<State>(new PhantomGoState(*this));
+}
+
+void PhantomGoState::UndoAction(Player player, Action action) {
+  // We don't have direct undo functionality, but copying the board and
```

Comment:
Observations should be popped as well, or maybe we should not implement this method at all.